cachi2/core/scm.py

```diff
@@ -1,16 +1,81 @@
 # SPDX-License-Identifier: GPL-3.0-or-later
 import logging
+import re
 import tarfile
 import tempfile
+import urllib.parse
+from os import PathLike
 from pathlib import Path
+from typing import NamedTuple, Union
 
-import git
+from git.repo import Repo
 
-from cachi2.core.errors import FetchError
+from cachi2.core.errors import FetchError, UnsupportedFeature
 
 log = logging.getLogger(__name__)
 
 
+class RepoID(NamedTuple):
+    """The properties which uniquely identify a repository at a specific commit."""
+
+    origin_url: str
+    commit_id: str
+
+    @property
+    def parsed_origin_url(self) -> urllib.parse.SplitResult:
+        """Get the url as a urllib.parse.SplitResult."""
+        return urllib.parse.urlsplit(self.origin_url)
+
+    def as_vcs_url_qualifier(self) -> str:
+        """Turn this RepoID into a 'vcs_url' qualifier as defined by the purl spec.
+
+        See https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#known-qualifiers-keyvalue-pairs
+        """
+        return f"git+{self.origin_url}@{self.commit_id}"
+
+
+def get_repo_id(repo: Union[str, PathLike[str], Repo]) -> RepoID:
+    """Get the RepoID for a git.Repo object or a git directory.
+
+    If the remote url is an scp-style [user@]host:path, convert it into ssh://[user@]host/path.
+
+    See `man git-clone` (GIT URLS) for some of the url formats that git supports.
+    """
+    if isinstance(repo, (str, PathLike)):
+        repo = Repo(repo)
+
+    try:
+        origin = repo.remote("origin")
+    except ValueError:
+        raise UnsupportedFeature(
+            "Cachi2 cannot process repositories that don't have an 'origin' remote",
+            solution=(
+                "Repositories cloned via git clone should always have one.\n"
+                "Otherwise, please `git remote add origin` with a url that reflects the origin."
+            ),
+        )
+
+    url = _canonicalize_origin_url(origin.url)
+    commit_id = repo.head.commit.hexsha
+    return RepoID(url, commit_id)
+
+
+def _canonicalize_origin_url(url: str) -> str:
+    if "://" in url:
+        return url
+    elif re.match("^[^/]*:", url):
+        # scp-style is "only recognized if there are no slashes before the first colon"
```

Comment:
Not sure if it's just me, but this comment would probably be more helpful if it was on top of the `elif`. 
(I tested the regex to figure out what it did, only to realize this comment explained it later :face_exhaling: )